File: DSA/Java/test/dsa/lib/algo/sort/RadixSortTest.java
package dsa.lib.algo.sort;

import org.junit.Test;

import static org.junit.Assert.assertArrayEquals;

public class RadixSortTest {
    @Test
    public void emptyArray() {
        assertRadixSort(new int[]{}, new int[]{});
    }

    @Test
    public void oneElement() {
        assertRadixSort(new int[]{1}, new int[]{1});
    }

    @Test
    public void sorts() {
        assertRadixSort(new int[]{1, 23, 45, 121, 432, 564, 788}, new int[]{121, 432, 564, 23, 1, 45, 788});
    }

    private void assertRadixSort(int[] expected, int[] nums) {
        radixSort(nums);
        assertArrayEquals(expected, nums);
    }

    private static void radixSort(int[] nums) {
        RadixSort.sort(nums);
    }
}

File: DSA/Cpp/test/dsa/lib/algo/sort/radix_sort_test.cpp
#include <vector>
#include <list>
#include <forward_list>

#include <dsa/testing/dsatest.h>
#include <dsa/lib/algo/sort/radix_sort.hpp>

namespace dsa::lib::algo::sort::rs {
namespace {

void expectRadixSort(std::vector<int> &&expected, std::vector<int> &&nums) {
    rs::sort(nums);
    EXPECT_EQ(expected, nums);
}
TEST(RadixSortTest, emptyArray) {
    expectRadixSort({}, {});
}
TEST(RadixSortTest, oneElement) {
    expectRadixSort({1}, {1});
}
TEST(RadixSortTest, sorts) {
    expectRadixSort({1, 23, 45, 121, 432, 564, 788}, {121, 432, 564, 23, 1, 45, 788});
}
TEST(RadixSortTest, iterator) {
    std::vector<int> nums = {121, 432, 564, 23, 1, 45, 788};
    std::vector<int> expected = {1, 23, 45, 121, 432, 564, 788};
    rs::sort(nums.begin(), nums.end());
    EXPECT_EQ(expected, nums);
}

void expectRadixSortStr(std::vector<std::string> &&expected, std::vector<std::string> &&ss) {
    rs::sort(ss);
    EXPECT_EQ(expected, ss);
}
TEST(RadixSortStrTest, emptyArray) {
    expectRadixSortStr({}, {});
}
TEST(RadixSortStrTest, oneElement) {
    expectRadixSortStr({"ab"}, {"ab"});
}
TEST(RadixSortStrTest, sorts) {
    expectRadixSortStr({"hac", "hke", "hzg", "iba", "ikf"}, {"hke", "iba", "hzg", "ikf", "hac"});
}
TEST(RadixSortStrTest, iterator) {
    std::vector<std::string> ss = {"hke", "iba", "hzg", "ikf", "hac"};
    std::vector<std::string> expected = {"hac", "hke", "hzg", "iba", "ikf"};
    rs::sortStr(ss.begin(), ss.end());
    EXPECT_EQ(expected, ss);
}

}
}
